src/devices/funkbus.c

```diff
@@ -0,0 +1,197 @@
+
+    funkbus_action_t action : 2;
+    uint8_t repeat : 1; // 1 == not first send of packet
+    uint8_t global : 1; // 1 == global ON / OFF button
+    uint8_t parity : 1; // parity over all bits
+    uint8_t check : 4;  // how to calculate is unknown
+} __attribute__((packed)) funkbus_packet_t;
+
+static uint64_t get_data_lsb(uint8_t const *bitrow, size_t start, uint8_t end)
+{
+    uint64_t result = 0;
+    uint64_t mask   = 1;
+    result          = 0;
+    for (; start <= end; mask <<= 1)
+        if (bitrow_get_bit(bitrow, start++) != 0)
+            result |= mask;
+    return result;
+}
+
+static uint8_t calc_parity(uint8_t const *bitrow, size_t len)
+{
+    uint8_t result = 0;
+    for (uint8_t i = 0; i < len; i++) {
+        if (bitrow_get_bit(bitrow, i)) {
+            result++;
+        }
+    }
+    return result & 0x01;
+}
+
+static int funkbus_callback(r_device *decoder, bitbuffer_t *bitbuffer)
+{
+
+    for (uint16_t row = 0; row < bitbuffer->num_rows; row++) {
+        uint16_t len = bitbuffer->bits_per_row[row];
+        uint8_t *bin  = bitbuffer->bb[row];
+
+        if (len < 48) {
+            return DECODE_ABORT_LENGTH;
+        }
+
+        uint8_t get_c = 0;
+#define get(x) \
+    get_data_lsb(bin, get_c, get_c + x - 1); \
+    get_c += x
+
+        funkbus_packet_t packet;
+        packet.typ    = get(4);
+        packet.subtyp = get(4);
+
+        // remote
+        if (packet.typ != 0x4 || packet.subtyp != 0x3) {
+            return DECODE_ABORT_EARLY;
+        }
+
+        packet.sn     = get(20);
+        packet.r1     = get(2);
+        packet.bat    = get(1);
+        packet.r2     = get(2);
+        packet.sw     = get(3);
+        packet.ch     = get(2);
+        packet.r3     = get(1);
+        packet.action = get(2);
+        packet.repeat = get(1);
+        packet.global = get(1);
+        packet.parity = get(1);
+        packet.check  = get(4); // how to calc?
+
+        if (packet.parity != calc_parity(bin, 43)) {
+            return DECODE_FAIL_MIC;
+        }
+
+        // calc id based on sn, ch + sw
+        uint64_t id = packet.sw + (packet.ch << 4) + (packet.sn << 8);
+
+        /* clang-format off */
+        data_t *data = data_make(
+                "model",        "",                DATA_STRING, "funkbus",
+                "typ",          "",                DATA_STRING, "remote",
+                "id",           "id",              DATA_INT, id,
+                "sn",           "serial number",   DATA_INT, packet.sn,
+                "bat",          "battery empty",   DATA_INT, packet.bat,
+                "sw",           "switch",          DATA_INT, packet.sw,
+                "ch",           "channel",         DATA_INT, packet.ch,
+                "action",       "action",          DATA_INT, packet.action,
+                "repeat",       "repeat",          DATA_INT, packet.repeat,
+                "global",       "global",          DATA_INT, packet.global,
+#ifdef DEBUG
```

Comment:
Change the conditional to a single block of `if (decoder->verbose > 1)` + fprintf

Reply:
no if since we are in the `data_make` function call ;) 
debug removed

Reply:
true. Generally it's fine to have fprintf (or one of the bitrow_*f / bitbuffer_printf) with error/debug data before the data output, given the decoder is set to verbose 1 or 2. E.g. `-R 202:vv` to set that at runtime.